File: Day2/solution.hs
import Data.List.Split (splitOn)

data Weapon = Rock | Paper | Scissor
  deriving Eq


instance Enum Weapon where
  toEnum n = case n of
    1 -> Rock
    2 -> Paper
    3 -> Scissor
    _ -> error "Invalid weapon"

  fromEnum Rock = 1
  fromEnum Paper = 2
  fromEnum Scissor = 3

  succ w = toEnum . max 1 $ (fromEnum w + 1) `mod` 4
  pred w = let x = fromEnum w - 1 `mod` 4 in if x == 0 then Scissor else toEnum x


toWeapon :: String -> Weapon
toWeapon s
  | s == "A" || s == "X" = Rock
  | s == "B" || s == "Y" = Paper
  | s == "C" || s == "Z" = Scissor
  | otherwise            = error "Invalid enum string"


input :: String -> IO [[Weapon]]
input name = map (map toWeapon . splitOn " ") . filter (/= "") . splitOn "\n" <$> readFile name


partOne :: [[Weapon]] -> Int
partOne = sum . map points
  where
    points :: [Weapon] -> Int
    points [o, s]
      | succ s == o = fromEnum s -- loss
      | s == o      = fromEnum s + 3 -- draw
      | otherwise   = fromEnum s + 6 -- win
    points _ = error "Invalid duel"


partTwo :: [[Weapon]] -> Int
partTwo = sum . map points
  where
    points :: [Weapon] -> Int
    points [o, s]
      | s == Rock = fromEnum $ pred o
      | s == Paper = 3 + fromEnum o
      | otherwise = 6 + fromEnum (succ o)
    points _ = error "Invalid duel"


main :: IO ()
main = do
  f <- input "input.txt"
  putStrLn $ "Part one: " ++ show (partOne f)
  putStrLn $ "Part two: " ++ show (partTwo f)

File: Day2/solution.cpp
#include <fstream>
#include <filesystem>
#include <iostream>

namespace stdfs = std::filesystem;

int part_one(stdfs::path const& path) {
    std::ifstream reader(path);
    int sum = 0;

    for (std::string line; std::getline(reader, line);) {
        char opponent = line[0];
        char self = line[2];

        // Won
        if ((self - opponent) % 3 == 0) { // self
            sum += 6;
        }
        // Draw
        else if (opponent == (self - 23)) {
            sum += 3;
        }

        // Rock = 1, Paper = 2, Scissors = 3
        sum += self - 87; // this effectively maps X = 1, Y = 2, Z = 3
    }

    return sum;
}

int part_two(stdfs::path const& path) {
    std::ifstream reader(path);
    int sum = 0;

    for (std::string line; std::getline(reader, line);) {
        char opponent = line[0] - 'A' + 1;
        char self = line[2];

        switch (self) {
            // Need a loss
            case 'X': {
                int loss = (opponent - 1) % 4;
                sum += loss == 0 ? 3 : loss; // handles the wrap around properly instead of going to 0
                break;
            }

            // Need a draw
            case 'Y': {
                sum += opponent + 3;
                break;
            }

            // Need a win
            case 'Z': {
                sum += std::max(1, (opponent + 1) % 4) + 6;
                break;
            }

            default:
                break;
        }
    }

    return sum;
}

int main() {
    auto filepath = stdfs::current_path().parent_path() / "input.txt";
    std::cout << "Part one: " << part_one(filepath) << '\n';
    std::cout << "Part two: " << part_two(filepath) << '\n';
}